### DISCUSSION TITLE
Solve SvelteKit Limitations for performance and i18n

### DISCUSSION THREAD

**Discussion by JaviKeziah** [sveltejs/kit#12001] [Ideas]
I've been using SvelteKit for several websites and have encountered some limitations that seem improvable. These limitations might be among the reasons why I'm not aware of any large companies using it in production.

All of these concerns are related to the app.html file. The inability to control your own head section results in performance issues, reduced control over your application, and difficulty implementing simple things.

For example:

- There is limited control over the order of the <head> elements, which is important for performance.

- When attempting to preload an image or add a `preconnect` to load images from a CDN, I've observed delays sometimes exceeding 0.5 seconds more than usual.

- It's not easy to place the Title at the beginning of the head where it is recommended (you can do it for every route with `transformPageChunk `, but it becomes tedious as your app scales).

- For internationalization (i18n), serving static pages without using a library or modifying the hooks seems unfeasible, as changing the %lang% attribute in `hooks.server.ts` is necessary.

- When using client-side routing, it's not a good idea to add a component inside <svelte:head> because it re-renders every time you navigate to the same page.

- Controlling the body tag for each specific route is also more difficult, as it is not scoped.

I believe these and other issues could be resolved by allowing the addition of Layouts as components with all the head elements replacing app.html file, similar to what is done in Astro or Next. This also would enable us to place `%sveltekit.head%` and `%sveltekit.body%` wherever needed for each route, facilitating easy implementation of internationalization without modifying `hooks.server.ts` and allowing a choice between static and dynamic routes with zero configuration.

Am I the only one thinking this @Rich-Harris?

I may be mistaken about some of these points, so if anyone knows an efficient way to address these challenges, I would love to hear it.

**Comment by teemingc** [comment]
cc: @benmccann has done a lot of investigation on SvelteKit's performance. I think he'll have some good ideas to share.

**Comment by benmccann** [comment]
> I believe these and other issues could be resolved by allowing the addition of Layouts as components with all the head elements replacing app.html file, similar to what is done in Astro or Next.

Do you have a link showing what exactly you're proposing? I don't have experience with those frameworks

**Reply by JaviKeziah** [reply]
### **Proposal 1**

The core of this proposal is to allow a file like the following `App.svelte` to act as a direct substitute for the traditional `app.html`. This file would serve as a dynamic component wrapper for a single route or for the `+layout.svelte`, so it could wrap multiple routes . Here's a conceptual example:

```
<script lang="ts">
        // src/app/App.svelte 

	 import SeoMetaTags from '../components/SeoMetaTags.svelte';
	 import ImgsToPreload from '../components/ImgsToPreload.svelte';

	 export let title = ""
</script>

<!doctype html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />	
		<meta name="viewport" content="width=device-width" />
		<title>{title}</title>
		<ImgsToPreload />
		<SeoMetaTags />

		%sveltekit.head% 

		<link rel="icon" type="image/svg+xml" href="/favicon.svg" />
	</head>
	<body data-sveltekit-preload-data="hover">
		<div style="display: contents">

%sveltekit.body%

</div>
	</body>
</html>
```

Key differences would be:

- This file would act as both a component and a wrapper for `+layout.svelte` or `+page.svelte` files.

- Developers could create multiple files like this, adhering to guidelines like including `%sveltekit.body%` within a specific place or tag.

--------------------------

### **Proposal 2 **

Adopting the same foundational approach as the first option but going a step further by eliminating the special `%sveltekit.head%` and `%sveltekit.body%` placeholders.  

The idea is constructing routes exclusively with components. This methodology echoes Astro's. This could demand more extensive modifications, which is a primary reason it was not my initial recommendation.

--------------------------


I believe both options would significantly simplify the customization of internationalization, offer developers complete control over the `<head>` tag without requiring prior knowledge of SvelteKit, and avoid many common issues that they might experience with the current configuration.

These are just some suggestions as a SvelteKit Developer who is using it in production for "scalable projects". There could be better alternatives that I am not aware of. 

What do you think?
